### PR TITLE
Track exports in exception tracing modules

### PR DESCRIPTION
**Pull request by robrix**
- [x] Depends on #668.
- [x] Generalizes functions & applications to 0-arity.
- [x] Adds exports to the module.
- [x] Connects the exports to their respective exception sets.

### CODE REVIEW

File: semantic-analysis/src/Analysis/Analysis/Typecheck.hs [hunk 1]
```diff
@@ -54,7 +54,7 @@ data Monotype a
   | Bool
   | Int
   | String
-  | Monotype a :-> Monotype a
+  | [Monotype a] :-> Monotype a
```
Comment (robrix) [inline semantic-analysis/src/Analysis/Analysis/Typecheck.hs:57]: The change to adopt n-ary functions is present in the domain interface, so all domains have to implement it. Thus, we extend our definition of function monotypes as well.

File: semantic-analysis/src/Analysis/Effect/Domain.hs [hunk 2]
```diff
@@ -39,10 +39,10 @@ dvar = send . DVar
 
 -- Functions
 
-dabs :: Has (Dom val) sig m => Name -> (val -> m val) -> m val
+dabs :: Has (Dom val) sig m => [Name] -> ([val] -> m val) -> m val
 dabs n = send . DAbs n
 
-dapp :: Has (Dom val) sig m => val -> val -> m val
+dapp :: Has (Dom val) sig m => val -> [val] -> m val
```
Comment (robrix) [inline semantic-analysis/src/Analysis/Effect/Domain.hs:45]: As described above, this is the key place where n-ary functions/applications are required.

File: semantic-analysis/src/Analysis/Syntax.hs [hunk 3]
```diff
@@ -79,6 +82,8 @@ eval eval = \case
     v' <- eval v
     let' n v' (eval b)
   Import ns -> S.simport ns >> dunit
+  Function n ps b -> letrec n (dabs ps (\ as ->
+    foldr (\ (p, a) m -> let' p a m) (eval b) (zip ps as)))
```
Comment (robrix) [inline semantic-analysis/src/Analysis/Syntax.hs:86]: Functions are treated as definitions (i.e. not lambdas), so we bind them recursively. In future we'll want to distinguish between definitions and their values, and between recursive and non-recursive bindings (i.e. `letrec` and `let`), but here again, it's good enough for demo work.

File: semantic-analysis/src/Analysis/Syntax.hs [hunk 4]
```diff
@@ -102,26 +114,28 @@ parseFile path = do
     -- FIXME: this should use the span of the source file, not an empty span.
     Right (_, Just root) -> pure (File (Ref.fromPath (Path.absRel path)) root)
 
-parseGraph :: A.Value -> A.Parser (IntMap.IntMap Term, Maybe Term)
+newtype Graph = Graph { terms :: IntMap.IntMap Term }
```
Comment (robrix) [inline semantic-analysis/src/Analysis/Syntax.hs:117]: For clarity. This is the adjacency list for the syntax tree; technically it's always tree-shaped, but we have no guarantee of ordering on when the nodes are deserialized, and even if we had one, the general case—for arbitrary `tree-sitter-graph` programs—is not guaranteed to be acyclic. (Thus our use of `Graph -> Term` during the decoding process, below.)

File: semantic-analysis/src/Analysis/Syntax.hs [hunk 5]
```diff
@@ -102,26 +114,28 @@ parseFile path = do
...
+
+parseGraph :: A.Value -> A.Parser (Graph, Maybe Term)
 parseGraph = A.withArray "nodes" $ \ nodes -> do
-  (untied, First root) <- foldMap (\ (k, v, r) -> ([(k, v)], First r)) <$> traverse (A.withObject "node" parseNode) (V.toList nodes)
-  -- @untied@ is a list of key/value pairs, where the keys are graph node IDs and the values are functions from the final graph to the representations of said graph nodes. Likewise, @root@ is a function of the same variety, wrapped in a @Maybe@.
+  (untied, First root) <- fold <$> traverse parseNode (V.toList nodes)
+  -- @untied@ is an intmap, where the keys are graph node IDs and the values are functions from the final graph to the representations of said graph nodes. Likewise, @root@ is a function of the same variety, wrapped in a @Maybe@.
```
Comment (robrix) [inline semantic-analysis/src/Analysis/Syntax.hs:122]: I rearranged where some of the types are packed/unpacked in an early attempt to ingest function names without making identifiers out of them. I eventually decided to match the name node directly and grab its source text for the string. (🎩 @dcreager for the suggestion.)

**Review by robrix — COMMENTED** [review]
This PR adds exports to `ExcSet` modules, ingesting function syntax & extending our domain to n-ary functions and applications en route.

**Comment by robrix** [comment]
@github/semantic-code: Ready for review!

File: semantic-analysis/src/Analysis/Analysis/Typecheck.hs [hunk 6]
```diff
@@ -230,16 +230,16 @@ instance ( Alternative m
...
+      ty <- hdl (b args <$ ctx)
+      pure ((args :->) <$> ty)
     L (DApp f a) -> do
-      arg <- meta
+      args <- traverse (const meta) a
       ret <- meta
-      unify f (arg :-> ret)
-      unify a arg
+      unify f (args :-> ret)
+      sequenceA_ (zipWith unify a args)
```
[the remark this answers is not in the document]
Comment (patrickt) [reply semantic-analysis/src/Analysis/Analysis/Typecheck.hs:242]: 🌎 

👨🏻‍🚀: _It’s all traverse?_

👨🏻‍🚀: 🔫  _Always has been._

File: semantic-analysis/src/Analysis/Syntax.hs [hunk 7]
```diff
@@ -102,26 +114,28 @@ parseFile path = do
...
-  -- We define @tied@ as the fixpoint of the former to yield the former as a graph of type @IntMap.IntMap Term@, and apply the latter to said graph to yield the entry point, if any, from which to evaluate.
-  let tied = fix (\ tied -> ($ tied) <$> IntMap.fromList untied)
-  pure (tied, ($ tied) <$> root)
+  -- We define @tied@ as the fixpoint of the former to yield the former as a graph of type @Graph@, and apply the latter to said graph to yield the entry point, if any, from which to evaluate.
+  let tied = fix (\ tied -> ($ Graph tied) <$> untied)
+  pure (Graph tied, ($ Graph tied) <$> root)
 
-parseNode :: A.Object -> A.Parser (IntMap.Key, IntMap.IntMap Term -> Term, Maybe (IntMap.IntMap Term -> Term))
-parseNode o = do
+parseNode :: A.Value -> A.Parser (IntMap.IntMap (Graph -> Term), First (Graph -> Term))
```
Comment (patrickt) [inline semantic-analysis/src/Analysis/Syntax.hs:128]: Can we get a description of what these return parameters are, and how they’re linked to those of `parseGraph`?

**Review by patrickt — APPROVED** [review]
Looks good; just got the one concern.

**Comment by robrix** [comment]
@patrickt: back to you 🙇🏻‍♂️

**Review by patrickt — APPROVED** [review]

**Pull request merged**